### PR TITLE
Fix old-ish regression on paypal standard self-service cancel links

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
This fixes what I suspect is an oldish regression whereby the cancel link for self-service paypal users is the CiviCRM internal page not the paypal page.

Before
----------------------------------------
- Cancel link points to civicrm cancel page for self-service paypal user. this is incorrect it should point to paypal
- Edit link present incorrectly for self-service paypal user. This is not offered & editing the recur record will not edit the recur
- Update billing link present & incorrect for paypal self service user
- view & cancel are present for manual processor recur
- all the same links show for an admin user viewing another contact's dashboard

image

After
----------------------------------------
1) no change for admin user viewing another contact's dashboard
1) the  cancel link CORRECTLY goes to the paypal website for the self service user
1) the edit link has CORRECTLY disappeared for the self service user
2) the update billing link is wrong but unchanged for the self service user

image

Technical Details
----------------------------------------
I've tried to keep the scope of this as narrow as possible while still addressing the actual bug identified in #18787 and trying to make at least some progress on 2 of the issues that were blocking #18196 and #18787- ie 

- lack of tests (this provides at least some protection and a basis for further PRs to extend as we seek to break free of the whack-a-mole previous fixes in this area have triggered) and
- some separation of the functionality for getting self-service vs administrative links

I think there is still quite a jumble in there around the back office and self-service being munged but if we ensure that as we tighten it up we keep adding tests we will be OK

Comments
----------------------------------------

### CODE REVIEW

**Comment by jitendrapurohit** [comment]
Hi @eileenmcnaughton, 

>the cancel link CORRECTLY goes to the paypal website for the self service user

As this is now redirecting to an external link, do we need to add no-popup to the Cancel link for PayPal standard? Currently, it opens as this by default.

image

>the edit link has CORRECTLY disappeared for the self service user

I do see an edit link on expanding the more option. Is that intentional? 

image

Probably due to the merge done at this point which seems like a required thing to do? https://github.com/civicrm/civicrm-core/pull/20210/files#diff-7117b779a415ec834465eaeffeaddfd50a003a8dab1b11192133adc264960727R165

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit you are totally spot on re the no popup class - I've added it.

Re the edit link - I think that is correct. You will still see an edit link if you have permission to edit the recurring contribution based on your civicrm permissions as a backoffice use - but less permissioned users should not.

Note if you re-do the r-run can you pull this PR in https://github.com/civicrm/civicrm-core/pull/20211 (which is a no change PR) so you test both together

File: tests/phpunit/CRM/Contribute/Page/TabTest.php [hunk 1]
```diff
@@ -18,22 +18,79 @@
...
+    $this->assertEquals($expected, $this->getActionHtml());
+
+    $page = new CRM_Contribute_Page_UserDashboard();
+    $page->run();
+    $expected = '<span><a href="https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_subscr-find&alias=sunil._1183377782_biz_api1.webaccess.co.in" class="action-item crm-hover-button" title=\'Cancel\' >Cancel</a>'
```
Comment (jitendrapurohit) [inline tests/phpunit/CRM/Contribute/Page/TabTest.php:63]: @eileenmcnaughton test failure seems to be due to the missing class in this line. https://test.civicrm.org/job/CiviCRM-Core-PR/40862/testReport/junit/(root)/CRM_Contribute_Page_TabTest/testLinksPaypalStandard/

**Comment by eileenmcnaughton** [comment]
Hopefully that fixes it (go jenkins go)

File: CRM/Contribute/Page/Tab.php [hunk 2]
```diff
@@ -94,6 +94,87 @@ public static function recurLinks(int $recurID, $context = 'contribution') {
...
+    if ($paymentProcessorObj->supports('cancelRecurring')
+      && $paymentProcessorObj->subscriptionURL($recurID, 'recur', 'cancel')
+    ) {
+      $url = $paymentProcessorObj->subscriptionURL($recurID, 'recur', 'cancel');
+      $links[CRM_Core_Action::DISABLE] = [
+        'url' => $url,
+        'name' => ts('Cancel'),
+        'title' => ts('Cancel'),
+        // Only display on-site links in a popup.
+        'class' => (stripos($url, 'http') !== FALSE) ? 'no-popup' : '',
```
Comment (jitendrapurohit) [inline CRM/Contribute/Page/Tab.php:121]: Hmm, looks like on-site links (eg dummy processor) get this class as well due to the absolute URL returned by `subscriptionURL()` in the base file - https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Payment.php#L1768.

**Comment by jitendrapurohit** [comment]
Thanks @eileenmcnaughton for the update. This looks fine to me. I've added a minor comment re the same `no-popup` class which i think requires a fix in the CRM/Core/Payment.php itself. +1 to merge 👍

**Review by jitendrapurohit — APPROVED** [review]

**Comment by eileenmcnaughton** [comment]
thanks @jitendrapurohit  - @seamuslee001 can you merge?

**Comment by seamuslee001** [comment]
Merging as per review from @jitendrapurohit

**Pull request merged**